File: go/data_generator.go
package main

import (
    "encoding/csv"
    "fmt"
    "math/rand"
    "os"
    "time"
)

func GenerateCSV(path string, rows int) error {
    file, err := os.Create(path)
    if err != nil {
        return err
    }
    defer file.Close()

    writer := csv.NewWriter(file)
    defer writer.Flush()

    // Write header
    header := []string{"Date", "Amount", "Account", "Department"}
    writer.Write(header)

    startDate, _ := time.Parse("2006-01-02", "2020-01-01")
    accounts := []string{"Sales", "Expenses", "Assets", "Liabilities"}
    departments := []string{"Sales", "Development", "HR", "Finance"}

    rand.Seed(time.Now().UnixNano())
    for i := 0; i < rows; i++ {
        offset := rand.Intn(365)
        date := startDate.AddDate(0, 0, offset).Format("2006-01-02")
        amount := rand.Float64()*990 + 10 // random float between 10 and 1000
        account := accounts[rand.Intn(len(accounts))]
        department := departments[rand.Intn(len(departments))]

        record := []string{
            date,
            fmt.Sprintf("%.2f", amount),
            account,
            department,
        }
        writer.Write(record)
    }
    return nil
}

File: rust/src/data_generator.rs
use chrono::{Duration, NaiveDate};
use rand::Rng;
use std::error::Error;
use std::fs::File;
use std::io::Write;

pub fn generate_csv(path: &str, rows: usize) -> Result<(), Box<dyn Error>> {
    let mut file = File::create(path)?;
    writeln!(file, "Date,Amount,Account,Department")?;

    let start_date = NaiveDate::from_ymd(2020, 1, 1);
    let accounts = vec!["Sales", "Expenses", "Assets", "Liabilities"];
    let departments = vec!["Sales", "Development", "HR", "Finance"];

    let mut rng = rand::thread_rng();

    for _ in 0..rows {
        let offset = rng.gen_range(0..365);
        let date = start_date + Duration::days(offset);
        let amount: f64 = rng.gen_range(10.0..1000.0);
        let account = accounts[rng.gen_range(0..accounts.len())];
        let department = departments[rng.gen_range(0..departments.len())];

        writeln!(
            file,
            "{},{},{},{}",
            date.format("%Y-%m-%d"),
            amount,
            account,
            department
        )?;
    }
    Ok(())
}
